### ISSUE TITLE
Use TensorPrimitives for better performance

### ISSUE THREAD

**Issue by sinshu** [sinshu/matflat#1]

**Comment by sinshu** [comment]
I replaced the implementation of several functions for `double` types with `TensorPrimitives` and benchmarked them.

* LU and Cholesky showed almost no change.
* QR and SVD became faster for large matrices.

It seems like introducing `TensorPrimitives` is worth considering, but we need to think about the following:

* Optimization for small matrices.
* Currently, the procedure to make `TensorPrimitives` available is too cumbersome 🥺

![image](https://github.com/user-attachments/assets/712d0f99-7df9-41a1-97b4-a30fdf78d2a8)

**Comment by MarsRover75** [comment]
Is the code branch that you used for testing available?
I was actually trying to do the same but the other way around, my point was to start with NET9 Tensors/TensorPrimitives namespace for basic operations and use NumFlat "interop" for something that's not covered or less effective/flexible in `System.Numerics.Tensors` namespace.
Didn't have much time to work on it though in terms of Tensor/TensorSpan usage with NumFLat.
One of the factors was the lack of `Span<T>` support as arguments (including out vars)  in NumFlat.

**Comment by sinshu** [comment]
Here is the branch:
https://github.com/sinshu/matflat/tree/net9

Regarding `Span<T>` support, to accept `Span`, the matrix would need to be a `ref struct`, but since `ref struct` cannot be placed on the heap, it posed practical usability issues. I also considered creating a type like `RefMat<T>` in addition to `Mat<T>`, but I gave up on the idea because it would complicate the API to a level I wouldn't be able to manage.

As a side note, the decision to use `Memory<T>` for the memory of `Mat<T>` was made after much deliberation. However, there are times when I wonder if raw arrays or `ArraySegment` might have been better. Accessing individual elements in `Memory<T>` is slower, after all...

**Comment by MarsRover75** [comment]
> The matrix would need to be a `ref struct`

No, the matrix doesn't need to be a ref struct itself, span is just a view of whatever data I have. It should be clients responsibility to allocate data and manage its lifecycle properly in high-performance scenarios, when it's possible.

For example, the library has many (basic!) methods that take `System.Collections.IEnumerable<T>` as an argument (bang, we're forced to use heap allocated collection or custom IEnumerable implementation, still heap-based), but doesn't take `Span<T>` or `ReadOnlySpan<T>`. 
Say I have my data as a `Span<T>` or `TensorSpan<T>` on a stack, what do I do to pass it to those library methods?

**Comment by MarsRover75** [comment]
> As a side note, the decision to use `Memory<T>` for the memory of `Mat<T>` was made after much deliberation. However, there are times when I wonder if raw arrays or `ArraySegment` might have been better. Accessing individual elements in `Memory<T>` is slower, after all...

Accessing individual elements in `Memory<T>` should be done using it's `Span` property (or it's slice), like:
```
var MySpan = MyMemory.Span;
// do whatever I need with elements
MySpan[i] ... 
```
this way it's pretty fast.

But the default [implementation](https://source.dot.net/#System.Memory/System/Buffers/ArrayMemoryPool.ArrayMemoryPoolBuffer.cs,12) of `MemoryPool<T>` is in fact just a wrapper for `ArrayPool<T>.Shared`
